packages/react-server/src/ReactDOMServerFormatConfig.js

```diff
@@ -7,17 +7,367 @@
+import type {Destination} from 'react-server/src/ReactServerStreamConfig';
 
-export function formatChunkAsString(type: string, props: Object): string {
-  let str = '<' + type + '>';
-  if (typeof props.children === 'string') {
-    str += props.children;
+import {
+  writeChunk,
+  convertStringToBuffer,
+} from 'react-server/src/ReactServerStreamConfig';
+
+import invariant from 'shared/invariant';
+
+// Per response,
+export type ResponseState = {
+  sentCompleteSegmentFunction: boolean,
+  sentCompleteBoundaryFunction: boolean,
+  sentClientRenderFunction: boolean,
+};
+
+// Allows us to keep track of what we've already written so we can refer back to it.
+export function createResponseState(): ResponseState {
+  return {
+    sentCompleteSegmentFunction: false,
+    sentCompleteBoundaryFunction: false,
+    sentClientRenderFunction: false,
+  };
+}
+
+// This object is used to lazily reuse the ID of the first generated node, or assign one.
+// This is very specific to DOM where we can't assign an ID to.
+export type SuspenseBoundaryID = {
+  id: null | string,
+};
+
+export function createSuspenseBoundaryID(
+  responseState: ResponseState,
+): SuspenseBoundaryID {
+  return {id: null};
+}
+
+function encodeHTMLIDAttribute(value: string): string {
+  // TODO: This needs to be encoded for security purposes.
+  return value;
+}
+
+function encodeHTMLTextNode(text: string): string {
+  // TOOD: This needs to be encoded for security purposes.
+  return text;
+}
+
+export function pushTextInstance(
+  target: Array<Uint8Array>,
+  text: string,
+): void {
+  target.push(convertStringToBuffer(encodeHTMLTextNode(text)));
+}
+
+const startTag1 = convertStringToBuffer('<');
+const startTag2 = convertStringToBuffer('>');
+
+export function pushStartInstance(
+  target: Array<Uint8Array>,
+  type: string,
+  props: Object,
+): void {
+  // TODO: Figure out if it's self closing and everything else.
+  target.push(startTag1, convertStringToBuffer(type), startTag2);
+}
+
+const endTag1 = convertStringToBuffer('</');
+const endTag2 = convertStringToBuffer('>');
+
+export function pushEndInstance(
+  target: Array<Uint8Array>,
+  type: string,
+  props: Object,
+): void {
+  // TODO: Figure out if it was self closing.
+  target.push(endTag1, convertStringToBuffer(type), endTag2);
+}
+
+// Structural Nodes
+
+// A placeholder is a node inside a hidden partial tree that can be filled in later, but before
+// display. It's never visible to users.
+const placeholder1 = convertStringToBuffer('<span id="');
+const placeholder2 = convertStringToBuffer('P:');
+const placeholder3 = convertStringToBuffer('"></span>');
+export function writePlaceholder(
+  destination: Destination,
+  id: number,
+): boolean {
+  // TODO: This needs to be contextually aware and switch tag since not all parents allow for spans like
+  // <select> or <tbody>. E.g. suspending a component that renders a table row.
```

Comment:
Why not an inline comment rather than a `<span>` to avoid this problem?

Reply:
Yea